File: fe/src/models/Dish.ts
import API from "./Custom/API.ts";
import Validate from "./Custom/Validate.ts";
import { z } from "zod";

export const DishItem = z.object({
  id: z.number().optional(),
  name: z.string().min(1),
  description: z.string().min(1),
  category_id: z.number(),
  price: z.number(),
  unit_of_measure_id: z.number(),
  image: z.string().default(""),
  allergens_id: z.array(z.number()).default([]),
  characteristics_id: z.array(z.number()).default([]),
});

export interface DishItem {
  id?: number;
  name: string;
  description: string;
  category_id: number;
  price: number;
  unit_of_measure_id: number;
  image: string;
  allergens_id: number[];
  characteristics_id: number[];
}

class Dish extends API<DishItem> implements DishItem {
  id?: number;
  name!: string;
  description!: string;
  category_id!: number;
  price!: number;
  unit_of_measure_id!: number;
  image!: string;
  allergens_id!: number[];
  characteristics_id!: number[];

  constructor(public dishItem: DishItem) {
    super("dish");
    Object.assign(this, dishItem);
  }

  async create(): Promise<DishItem> {
    return super.create(this);
  }

  static async getAll(): Promise<DishItem[] | never> {
    return new API<DishItem>("dish").getAll();
  }

  static get(id: NonNullable<DishItem["id"]>): Promise<DishItem | never> {
    return new API<DishItem>("dish").get(id);
  }

  async delete(): Promise<DishItem> {
    if (!this.id) throw new Error("No id provided");
    return super.delete(this.id);
  }

  async update(): Promise<DishItem> {
    if (!this.id) throw new Error("No id provided");
    return super.update(this.id, this);
  }

  async uploadImage(image: File) {
    if (!this.id) throw new Error("No id provided");
    const formData = new FormData();
    formData.append("image", image);
    return this.axios.post<string>(`/image/${this.id}`, formData);
  }

  async deleteImage() {
    if (!this.id) throw new Error("No id provided");
    return this.axios.delete<string>(`/image/${this.id}`);
  }

  get imgUrl(): URL | undefined {
    if (!this.image || !this.id) return;
    return new URL(`${this.axios.defaults.baseURL}/image/${this.id}`);
  }

  static validate(obj: any): boolean {
    const dish = new Dish(obj);
    const result: any[] = [];
    result.push(Validate.string(dish.name));
    result.push(Validate.string(dish.description));
    result.push(Validate.number(dish.category_id));
    result.push(Validate.number(dish.price));
    result.push(Validate.number(dish.unit_of_measure_id));
    console.log(result);
    console.log(dish);
    return result.every((item) => item);
  }
}

export default Dish;

File: be/routes/Dish.py
import os

from flask import request, make_response, jsonify, send_from_directory

from .API import RouteMethodView
from .CRUD import CRUD
from .db import db


class Dish(db.Model):
    __tablename__ = 'dish'

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    description = db.Column(db.Text)
    category_id = db.Column(db.Integer, db.ForeignKey('category.id'))
    price = db.Column(db.Numeric(10, 2))
    unit_of_measure_id = db.Column(db.Integer, db.ForeignKey('unitofmeasure.id'))
    image = db.Column(db.String(255))
    allergens_id = db.Column(db.ARRAY(db.Integer), default=[])
    characteristics_id = db.Column(db.ARRAY(db.Integer), default=[])


class DishView(CRUD, RouteMethodView):
    route = Dish.__tablename__
    path = '/public/dish' if os.getenv('FLASK_ENV') == 'production' else os.path.join(os.getcwd(), 'dish')

    def __init__(self):
        super().__init__(Dish)

    def upload_image(self, id: int):
        if 'image' not in request.files:
            return make_response('Bad request', 400)
        file = request.files['image']
        filename = f'image_{id}' if file.filename != '' else ''
        if not filename:
            return make_response('Bad request', 400)

        try:
            if not os.path.exists(self.path):
                os.makedirs(self.path)
            file.save(os.path.join(self.path, filename))
            with self.app.app_context():
                dish_elem = self.db.session.query(Dish).filter_by(id=id).first()
                dish_elem.image = filename
                self.db.session.commit()
                return filename, 201
        except Exception as e:
            return make_response(jsonify({'result': False, 'error': str(e)}), 500)

    def serve_image(self, id: int):
        with self.app.app_context():
            dish_elem = self.db.session.query(Dish).filter_by(id=id).first()
            if not dish_elem:
                return make_response('Not found', 404)
            if not dish_elem.image:
                return make_response('', 204)
            return send_from_directory(self.path, dish_elem.image)

    def delete_image(self, id: int):
        with self.app.app_context():
            dish_elem = self.db.session.query(Dish).filter_by(id=id).first()
            if not dish_elem:
                return make_response('Not found', 404)
            os.remove(os.path.join(self.path, dish_elem.image))
            dish_elem.image = ''
            self.db.session.commit()
            return make_response('', 204)

    def image(self, id: int):
        method_func = {
            'POST': self.upload_image,
            'GET': self.serve_image,
            'DELETE': self.delete_image,
        }
        if request.method not in method_func:
            return make_response('Method not allowed', 405)
        return method_func[request.method](id)
    @property
    def custom_routes(self) -> dict[str, dict[str, any]]:
        return {
            'image/<int:id>': {
                'methods': ['POST', 'GET', 'DELETE'],
                'view_func': self.image,
            },
        }
